lib/json-schema/schema/reader.rb

```diff
@@ -3,19 +3,37 @@
 
 module JSON
   class Schema
-    # Raised by {JSON::Schema::Reader} when one of its settings indicate
-    # a schema should not be readed.
-    class ReadRefused < StandardError
+    # Base for any reading exceptions encountered by {JSON::Schema::Reader}
+    class ReadError < StandardError
       # @return [String] the requested schema location which was refused
       attr_reader :location
 
       # @return [Symbol] either +:uri+ or +:file+
       attr_reader :type
 
+      private
+
+      def type_string
+        type == :uri ? 'URI' : type.to_s
+      end
+    end
+
+    # Raised by {JSON::Schema::Reader} when one of its settings indicate
+    # a schema should not be readed.
+    class ReadRefused < ReadError
+      def initialize(location, type)
+        @location = location
+        @type = type
+        super("Read of #{type_string} at #{location} refused!")
+      end
+    end
+
+    # Raised by {JSON::Schema::Reader} when an attempt to read a schema fails
+    class ReadFailed < ReadError
       def initialize(location, type)
         @location = location
```

Comment:
It feels like the parent initializer should set location and type, as type string depends on it, and the code is shared between ReadRefused and ReadFailed